formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Encoding.kt

```diff
@@ -467,3 +573,22 @@ private fun Iterable<ByteArray>.flatten(): ByteArray {
 private fun SerialDescriptor.isByteString(index: Int): Boolean =
     getElementDescriptor(index) == ByteArraySerializer().descriptor &&
         getElementAnnotations(index).find { it is ByteString } != null
+
+/**
+ * Removes an item from the top of the [lengthStack], cascading the removal if the item represents the last item (i.e. a
+ * length value of `1`) at its stack depth.
+ *
+ * For example, pruning a [lengthStack] of `[3, 2, 1, 1]` would result in `[3, 1]`.
+ */
+private fun prune(lengthStack: MutableList<Int>) {
```

Comment:
Since this method is tied to a particular logic and algorithm (in contrary to `ByteArray.flatten()`), I'd suggest to move it inside decoder next to `skipElement` and `isIndefinite`